### PR TITLE
Implementation of single threaded downloader

### PR DESCRIPTION
**Pull request by Psyhich**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+
+add_executable(tests ${TESTS})
+target_link_libraries( tests
+	gtest # For basic library usage
+	gtest_main # For tests to be runned(this rung RUN_ALL_TESTS())
+	rt # Needed for gtest
+	pthread # Needed for gtest
+	${PROJECT_NAME} # My test files
+)
```
Comment (VictorApopiy) [inline CMakeLists.txt:64]: Let's Add gtest as a third party this time

File: include/sockets.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace Sockets
+{
+	/* 
+	* Representation of simple blocking socket that can send and recive data
+	*/
+	class CSocket 
+	{
+	public:
+		CSocket(const std::string& addressToUse);
```
Comment (VictorApopiy) [inline include/sockets.h]: use the correct prefix

File: include/sockets.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		virtual std::optional<std::vector<char>> ReadTillEnd() noexcept = 0;
+		virtual bool Write(const char* pcchBytes, size_t nCount)  noexcept = 0;
+		virtual ~CSocket() {}
+	protected:
+		inline const URI& GetAddress() { return m_address; }
```
Comment (VictorApopiy) [inline include/sockets.h]: this method could be const

File: include/uri.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#ifndef MY_URI_H
+#define MY_URI_H
+
+#include <string>
+#include <optional>
+
+class URI {
```
Comment (VictorApopiy) [inline include/uri.h]: use prefix C

File: include/sockets.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		bool Write(const char* pcchBytes, size_t nCount) noexcept override;
+
+		CTcpSocket(const CTcpSocket &cSockeToCopy) = delete;
+		CTcpSocket operator=(const CTcpSocket &cSockeToCopy) = delete;
+
+		CTcpSocket(CTcpSocket &&socketToMove);
+		CTcpSocket operator =(CTcpSocket &&socketToMove);
+	private:
+		// Return port in network byte order
+		static std::optional<uint16_t> ExtractPortInByteOrder(const URI &cURIString) noexcept;
```
Comment (VictorApopiy) [inline include/sockets.h]: are you sure that this function should be static and private?

File: src/sockets.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		const sockaddr cFirstAddress = *ppResolvedHosts[0].ai_addr;
+		freeaddrinfo(ppResolvedHosts);
+		return cFirstAddress;
+	}
+	return std::nullopt;
+}
+
+
+Sockets::CTcpSocket::CTcpSocket(
+	const std::string& addressToUse) : CSocket(addressToUse)
```
Comment (VictorApopiy) [inline src/sockets.cpp:56]: cs prefix

File: src/sockets.cpp [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	{
+		fprintf(stderr, "The given socket doesn't exist or it's not opened\n");
+		return std::nullopt;
+	}
+	
+	char chBuffer[BUFFER_SIZE];
+	memset(chBuffer, 0, sizeof(char) * BUFFER_SIZE);
+
+	std::vector<char> messageVector;
+	messageVector.reserve(BUFFER_SIZE * 4);
```
Comment (VictorApopiy) [inline src/sockets.cpp]: Why BUFFER_SIZE * 4?

File: src/sockets.cpp [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	// If we recive same count as BUFFER_SIZE, we can request new block of data
+	ssize_t nRecivedBytes;
+	do
+	{
+		nRecivedBytes = recv(m_iSocketFD, chBuffer, sizeof(char) * BUFFER_SIZE, 0);
+		if(nRecivedBytes == -1)
+		{
+			fprintf(stderr, "Error while reciving data\n");
+			return std::nullopt;
+		} else if(nRecivedBytes == 0)
```
Comment (VictorApopiy) [inline src/sockets.cpp]: new line

File: src/uri.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+		{
+			nProtocolLength = nIndex - 2;
+			break;
+		}
+	}
+
+	if(nProtocolLength == 0)
+	{
+		return std::nullopt;
+	} else
```
Comment (VictorApopiy) [inline src/uri.cpp]: new line

File: src/uri.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		const int ciParsedNumber = std::stoi(cStringPort.c_str());
+		if(ciParsedNumber > 65535) // This is larges port number
+		{
+			return std::nullopt;
+		}
+		return ciParsedNumber;
+	} catch(std::invalid_argument &err)
+	{
+		return std::nullopt;
+	}
```
Comment (VictorApopiy) [inline src/uri.cpp]: Is this method noexcept?

**Review by VictorApopiy — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (Psyhich) [reply CMakeLists.txt:64]: Fixed by adding git submodule in lib directory

File: include/uri.h (hunk 4 above)
Comment (Psyhich) [reply include/uri.h]: Fixed

File: include/sockets.h (hunk 2 above)
Comment (Psyhich) [reply include/sockets.h]: Fixed

File: include/sockets.h (hunk 3 above)
Comment (Psyhich) [reply include/sockets.h]: Changed to const

File: src/sockets.cpp (hunk 6 above)
Comment (Psyhich) [reply src/sockets.cpp:56]: Added prefix

File: src/sockets.cpp (hunk 7 above)
Comment (Psyhich) [reply src/sockets.cpp]: Removed "* 4" as it we can manually set BUFFER_SIZE

File: src/sockets.cpp (hunk 8 above)
Comment (Psyhich) [reply src/sockets.cpp]: Moved to new line

File: src/uri.cpp (hunk 9 above)
Comment (Psyhich) [reply src/uri.cpp]: Moved to new line

File: src/uri.cpp (hunk 10 above)
Comment (Psyhich) [reply src/uri.cpp]: Fixed by catching any std::exception because stoi throws invalid_argument or out_of_range

File: include/sockets.h (hunk 5 above)
Comment (Psyhich) [reply include/sockets.h]: Changed to method

**Comment by Psyhich** [comment]
Tests and better documentation will be in next pull request

**Pull request merged**